File: services/gateway/gateway-api/src/main/java/com/example/gateway/api/service/StockDataService.java
//package com.example.gateway.api.service;
//
//import com.example.gateway.api.converter.ShareDataConverter;
//import com.example.gateway.api.gateway.StockDataGateway;
//import com.example.stockdata.api.spec.model.PriceItemResponse;
//import org.springframework.stereotype.Service;
//
//@Service
//public class StockDataService {
//
//    private StockDataGateway stockDataGateway;
//
//    public StockDataService(StockDataGateway stockDataGateway) {
//        this.stockDataGateway = stockDataGateway;
//    }
//
//    public ShareDataGWModel getData(String accountId, String symbol) {
//        PriceItemResponse returnModel = stockDataGateway.getPriceDataBySymbol(accountId, symbol);
//        return ShareDataConverter.toShareDataGWModel.apply(returnModel);
//    }
//}

File: services/gateway/gateway-api/src/main/kotlin/com/example/gateway/api/core/service/StockDataService.kt
package com.example.gateway.api.core.service

import arrow.core.Either
import com.example.common.converter.TimeConverter
import com.example.common.exception.ApiException
import com.example.gateway.api.amqp.AmqpSender
import com.example.gateway.api.core.model.SharePriceModel
import com.example.gateway.api.rest.gateway.AlphaVantageGateway
import com.example.gateway.api.rest.gateway.StockDataGateway
import org.springframework.stereotype.Service

@Service
class StockDataService(
        private val stockDataGateway: StockDataGateway,
        private val alphaVantageGateway: AlphaVantageGateway,
        private val ampqSender: AmqpSender) {

    fun getPrice(symbol: String) = stockDataGateway.getPrice(symbol)

    fun updatePrice(model: SharePriceModel): Either<ApiException, Unit> {
        val now = TimeConverter.getOffsetDateTimeNow()

        return alphaVantageGateway
                .getQuoteResponse(model.symbol)
                .bimap(
                        {
                            var active = model.active
                            if (model.lastFailedOn != null) {
                                active = false
                            }

                            model.lastFailedOn = now
                            model.lastUpdatedOn = now
                            model.active = active
                            ampqSender.updatePrice(model)
                            it
                        },
                        {
                            it.lastUpdatedOn = now
                            ampqSender.updatePrice(it)
                        }
                )

    }

    fun getSymbolsForUpdate() = stockDataGateway.getSymbolsForUpdate()

    fun updateSymbols(symbols: List<String>) = stockDataGateway.updateSymbols(symbols)
}
